src/node_webstorage.cc

```diff
@@ -0,0 +1,706 @@
+      "  single_row_ INTEGER NOT NULL DEFAULT 1 CHECK(single_row_ = 1),"
+      "  PRIMARY KEY(single_row_)"
+      ") STRICT;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_insert "
+      "AFTER INSERT ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_state"
+      "    SET total_size = total_size + OCTET_LENGTH(NEW.key) +"
+      "      OCTET_LENGTH(NEW.value);"
+      "  SELECT RAISE(ABORT, 'QuotaExceeded') WHERE EXISTS ("
+      "    SELECT 1 FROM nodejs_webstorage_state WHERE total_size > max_size"
+      "  );"
+      "END;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_update "
+      "AFTER UPDATE ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_state"
+      "    SET total_size = total_size + "
+      "      ((OCTET_LENGTH(NEW.key) + OCTET_LENGTH(NEW.value)) -"
+      "      (OCTET_LENGTH(OLD.key) + OCTET_LENGTH(OLD.value)));"
+      "  SELECT RAISE(ABORT, 'QuotaExceeded') WHERE EXISTS ("
+      "    SELECT 1 FROM nodejs_webstorage_state WHERE total_size > max_size"
+      "  );"
+      "END;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_delete "
+      "AFTER DELETE ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_state"
+      "    SET total_size = total_size - (OCTET_LENGTH(OLD.key) +"
+      "      OCTET_LENGTH(OLD.value));"
+      "END;"
+      ""
+      "INSERT OR IGNORE INTO nodejs_webstorage_state (total_size) VALUES (0);";
+
+  sqlite3* db = db_.get();
+  if (db != nullptr) {
+    return true;
+  }
+
+  int r = sqlite3_open(location_.c_str(), &db);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, false);
+  r = sqlite3_exec(db, init_sql_v0, 0, 0, nullptr);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, false);
+
+  // Get the current schema version, used to determine schema migrations.
+  sqlite3_stmt* s = nullptr;
+  r = sqlite3_prepare_v2(db, get_schema_version_sql, -1, &s, 0);
+  r = sqlite3_exec(db, init_sql_v0, 0, 0, nullptr);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, false);
+  auto stmt = stmt_unique_ptr(s);
+  CHECK_ERROR_OR_THROW(env(), sqlite3_step(stmt.get()), SQLITE_ROW, false);
+  CHECK(sqlite3_column_type(stmt.get(), 0) == SQLITE_INTEGER);
+  int schema_version = sqlite3_column_int(stmt.get(), 0);
+  stmt = nullptr;  // Force finalization.
+
+  if (schema_version > kCurrentSchemaVersion) {
+    node::THROW_ERR_INVALID_STATE(
+        env(), "localStorage was created with a newer version of Node.js");
+    return false;
+  }
+
+  if (schema_version < kCurrentSchemaVersion) {
+    // Run any migrations and update the schema version.
+    std::string set_user_version_sql =
+        "UPDATE nodejs_webstorage_state SET schema_version = " +
+        std::to_string(kCurrentSchemaVersion) + ";";
+    r = sqlite3_exec(db, set_user_version_sql.c_str(), 0, 0, nullptr);
+    CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, false);
+  }
```

Comment:
I have a rough feeling we would need to implement a migration framework here. Is that risk mitigated somewhere?

Reply:
The idea is to use the schema version for managing any required migrations.